File: Java/src/Napakalaki/Napakalaki.java
package Napakalaki;
import java.util.*;

public class Napakalaki {
    // SINGLETON
    private static final Napakalaki instance = new Napakalaki();
    private Player currentPlayer;
    private ArrayList<Player> players;
    private Monster currentMonster; 
    
    private Napakalaki() {
        players = new ArrayList();
    }
    
    private void initPlayers( ArrayList<String> names ){
    
        for(int i= 0; i < names.size() ; i++){
            players.add( new Player(names.get(i)) );
        }
        
    }
    
    private Player nextPlayer()
    {
        int indice;
        
        if ( currentPlayer == null )
            
            indice = (int) ( Math.random() * players.size() + 0 );
        
        else
        {
             int indice_actual = players.indexOf( currentPlayer );
             
             if ( indice_actual == players.size()-1 )
                 indice = 0;
             else
                 indice = indice_actual + 1;
        }
        currentPlayer = players.get( indice );
        
        return currentPlayer;
    }
    
    public boolean nextTurnAllowed(){
        boolean valido = true;

        if (this.currentPlayer == null)
            valido = true;
        
        else
            valido = this.currentPlayer.validState();

        return valido;
    }
    private void setEnemies(){
        for (int i = 0; i < players.size(); i++){
            int actual;
            
            do {
                Random enemigo = new Random();
                actual = enemigo.nextInt(players.size());
            }while(actual == i);
            
            players.get(i).setEnemy( players.get(actual) );
        }
    }
    public static Napakalaki getInstance() {
        return instance;
    }
    public CombatResult developCombat()
    {
        CombatResult combat = currentPlayer.combat(currentMonster);
        CardDealer.getInstance().giveMonsterBack(currentMonster);
        
        if ( combat == CombatResult.LOSEANDCONVERT )
        {
            Cultist cultist_card = CardDealer.getInstance().nextCultist();
            CultistPlayer cultist_player = new CultistPlayer( currentPlayer, cultist_card );
            
            // Cambiamos el enemigo del resto de jugadores por el cultista
            for ( Player it : players ) 
            {
                if ( currentPlayer == it.getEnemy() )
                    it.setEnemy( cultist_player );
            }
            // Cambiamos el jugador actual por el cultista en la lista de jugadores.
            for ( Player it : players )
            {
                if ( it == currentPlayer )
                    it = cultist_player;
            }
            // Lo cambiamos en la variable jugador actual
            currentPlayer = cultist_player;
            
        }
        
        return combat;
    }
    public void discardVisibleTreasures( ArrayList<Treasure> treasures ) {
        for (Treasure treasure : treasures) {
            currentPlayer.discardVisibleTreasure(treasure);
            CardDealer.getInstance().giveTreasureBack(treasure);
        }
    }
    public void discardHiddenTreasures( ArrayList<Treasure> treasures ) {
        for(Treasure treasure : treasures) {
           currentPlayer.discardHiddenTreasure(treasure);
           CardDealer.getInstance().giveTreasureBack(treasure);
        } 
    }
    
    public void makeTreasuresVisible( ArrayList<Treasure> treasures ) {
        for ( Treasure t : treasures )
            currentPlayer.makeTreasureVisible( t );
    }
    
    public void initGame( ArrayList<String> players ) {
        initPlayers(players);
        setEnemies();
        CardDealer.getInstance().initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer() {
        return currentPlayer;
    }
    
    public Monster getCurrentMonster() {
        return currentMonster;
    }
    
    public boolean nextTurn() {
        boolean stateOK = nextTurnAllowed();
        if ( stateOK )
        {
            this.currentMonster = CardDealer.getInstance().nextMonster();
            this.currentPlayer =  nextPlayer();
            if( currentPlayer.isDead() ){
                currentPlayer.initTreasures();
            }
        }
        return stateOK;
    }
    
    public boolean endOfGame( CombatResult result ) {
        return ( result == CombatResult.WINGAME );
    }
    
}

File: ruby/lib/Napakalaki.rb
#encoding: UTF-8

require "singleton"
require_relative "BadConsequence"
require_relative "CardDealer"
require_relative "CombatResult"
require_relative "Dice"
require_relative "Monster"
require_relative "Player"
require_relative "Prize"
require_relative "Treasure"
require_relative "TreasureKind"

module NapakalakiGame
  
  class Napakalaki

    include Singleton
    attr_accessor :currentPlayer, :players, :dealer, :currentMonster

    def initialize
      @currentPlayer = Player.new( "" )
      premio = Prize.new( 0,0 )
      bc = BadConsequence.newLevelNumberOfTreasures( "", 0, 0, 0 )
      @currentMonster = Monster.new( 0, 0, premio, bc )
    end

    def initPlayers( names )
      @dealer = CardDealer.instance
      @players = Array.new
      names.each do |name|
        players << Player.new( name )
      end
    end
    private :initPlayers

    def nextPlayer
      # Si nunca se ha inizializado entonces empieza cualquier jugador 
      # aleatoriamente.
      if @current_player == nil then
        indice = rand( @players.size )

      # De lo contrario hay dos opciones
      else
        indice_actual = @players.index( @currentPlayer )

        # Que sea el ultimo jugador de la lista, por lo que empieza por el 
        # primero de nuevo.
        if indice_actual == @players.size then
          indice = 0
        # Que no sea el ultimo por lo que salta al siguiente.
        else
          indice = indice_actual + 1
        end
      end
      @current_player = @players[ indice ]

      @current_player
    end
    private :nextPlayer

    def nextTurnAllowed
      if @current_player == nil then
          permitido = true
      else
          permitido = @current_player.validState
      end
      permitido
    end
    private :nextTurnAllowed

    def setEnemies
      for i in 0..( @players.size - 1 )
        enemigo = rand( @players.size )
        # Hasta que no genere un numero aleatorio que no coincida con su indice
        # no sigue adelante.
        while i != enemigo do
          enemigo = rand( @players.size )
        end
        @players[i].enemy = @players[ enemigo ]
      end
    end
    private :setEnemies

    def developCombat
      @currentPlayer.combat( @currentMonster )
    end

    def discardVisibleTreasures( treasures )
      for i in 0..( treasures.size - 1 )
        @currentPlayer.discartVisibleTreasure( treasures[i] )
        @dealer.giveThreasureBack( treasures[i] )
      end
    end

    def discardHiddenTreasures( treasures )
      for i in 0..( treasures.size - 1 )
        @currentPlayer.discartHiddenTreasure( treasures[i] )
        @dealer.giveThreasureBack( treasures[i] )
      end
    end

    def makeTreasuresVisible( treasures )
      for i in 0..( treasures.size - 1 )
        @currentPlayer.makeTreasureVisible( treasures[i] )
      end
    end

    def initGame( players )
      initPlayers( players )
      setEnemies
      nextTurn
      @dealer.initCards
    end

    def nextTurn
      stateOK = nextTurnAllowed

      if stateOK then
        @currentMonster = @dealer.nextMonster
        @currentPlayer = nextPlayer
        dead = @currentPlayer.dead

        if dead then
          @currentPlayer.initTreasures
        end
      end
      stateOK
    end

    def endOfGame( result )
      return ( result == WINGAME )
    end

    def getCurrentPlayer
      @currentPlayer
    end

    def getCurrentMonster
      @currentMonster
    end
  end

end